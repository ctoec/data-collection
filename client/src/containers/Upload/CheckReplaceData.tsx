import React, { useContext, useState, useEffect } from 'react';
import qs from 'query-string';
import {
  Modal,
  CheckboxGroup,
  Checkbox,
  CheckboxOption,
  Button,
} from '@ctoec/component-library';
import UserContext from '../../contexts/UserContext/UserContext';

type CheckReplaceDataProps = {
  isOpen: boolean;
  toggleIsOpen: () => void;
  clearFile: () => void;
  setPostUpload: (_: boolean) => void;
  setQueryString: (_: string) => void;
};
export const CheckReplaceData: React.FC<CheckReplaceDataProps> = ({
  isOpen,
  toggleIsOpen,
  clearFile,
  setPostUpload,
  setQueryString,
}) => {
  const { user } = useContext(UserContext);
  const [siteIdsToReplace, setSiteIdsToReplace] = useState<number[]>([]);

  useEffect(() => {
    setQueryString(
      siteIdsToReplace.length
        ? `?${qs.stringify({ overwriteSites: siteIdsToReplace })}`
        : ''
    );
  }, [siteIdsToReplace]);

  const options: CheckboxOption[] = [
    // All sites option
    {
      value: 'all',
      render: (props) => (
        <Checkbox
          {...props}
          checked={siteIdsToReplace.length === 0}
          id="all-sites"
          text="All sites"
          onChange={(e) => {
            if (e.target.checked) {
              setSiteIdsToReplace([]);
            }
          }}
        />
      ),
    },
    ...(user?.sites || []).map(
      (site) =>
        ({
          value: `${site.id}`,
          render: (props) => (
            <Checkbox
              {...props}
              value={`${site.id}`}
              checked={siteIdsToReplace.includes(site.id)}
              id={`site-${site.id}`}
              text={site.siteName}
              onChange={(e) => {
                const siteId = parseInt(e.target.value);
                if (e.target.checked) {
                  setSiteIdsToReplace((ids) => [...ids, siteId]);
                } else {
                  setSiteIdsToReplace((ids) =>
                    ids.filter((id) => id !== siteId)
                  );
                }
              }}
            />
          ),
        } as CheckboxOption)
    ),
  ];

  return (
    <Modal
      isOpen={isOpen}
      toggleOpen={toggleIsOpen}
      header={<h2> Do you want to replace your existing data? </h2>}
      showHeaderBorder
      content={
        <div>
          <p>
            When this file is uploaded, all of your existing data will be
            overwritten.
          </p>
          <p>
            If this upload is only meant to replace data for specific sites,
            choose them below:
          </p>

          <CheckboxGroup
            id="replace-data-sites"
            legend="Choose sites"
            showLegend
            horizontal
            options={options}
          />

          <div className="margin-top-2">
            <Button
              appearance="outline"
              text="Cancel"
              onClick={() => {
                clearFile();
                toggleIsOpen();
              }}
            />
            <Button
              text="Replace data"
              onClick={() => {
                setPostUpload(true);
                toggleIsOpen();
              }}
            />
          </div>
        </div>
      }
    />
  );
};
