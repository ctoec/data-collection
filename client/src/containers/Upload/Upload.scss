.three-column-layout {
  display: flex;
  flex-wrap: wrap;
  flex-direction: row;
  flex: 1;
  align-items: flex-start;
}

.three-column-card {
  margin: 0.5em;
  width: 30%;
  max-width: 30%;
}
