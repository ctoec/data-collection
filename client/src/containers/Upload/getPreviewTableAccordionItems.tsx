import React from 'react';
import pluralize from 'pluralize';
import {
  AccordionItemProps,
  ErrorBoundary,
  InlineIcon,
  Table,
} from '@ctoec/component-library';
import { defaultErrorBoundaryProps } from '../../utils/defaultErrorBoundaryProps';
import { AgeGroup } from '../../shared/models';
import { getPreviewTableColumns } from './previewTableColumns';
import { UploadPreviewRow } from './UploadPreviewRow';

export const MAX_LENGTH_EXPANDED = 50;
export const QUERY_STRING_MONTH_FORMAT = 'MMMM-YYYY';

/**
 * Helper types for organizing children into sections by ageGroup
 */
export const NoAgeGroup = 'Incomplete enrollments';
export type RosterSections = AgeGroup | typeof NoAgeGroup;
export type RowsByAgeGroup = {
  [key in RosterSections]?: UploadPreviewRow[];
};

/**
 * Returns array of AccordionItemProps, used to render the preview
 * age group accordion sections.
 * @param children
 * @param opts
 */
export function getPreviewTableAccordionItems(
  children: UploadPreviewRow[]
): AccordionItemProps[] {
  const childrenByAgeGroup = getChildrenByAgeGroup(children);

  return Object.entries(childrenByAgeGroup)
    .filter(
      ([_, ageGroupChildren]) => ageGroupChildren && ageGroupChildren.length
    )
    .map(([ageGroup, ageGroupChildren = []]) => ({
      id: ageGroup.replace(' ', '-'),
      title: (
        <>
          {ageGroup === NoAgeGroup && <InlineIcon icon="attentionNeeded" />}
          {ageGroup}{' '}
          <span className="text-normal">
            {pluralize('children', ageGroupChildren.length, true)}{' '}
          </span>
          {ageGroup === NoAgeGroup && (
            <>
              <br />
              <p className="usa-error-message margin-y-0 font-body-md">
                All records must have an assigned age group
              </p>
            </>
          )}
        </>
      ),
      expandText: (
        <>
          Show<span className="usa-sr-only">{` ${ageGroup} roster`}</span>
        </>
      ),
      collapseText: (
        <>
          Hide<span className="usa-sr-only">{` ${ageGroup} roster`}</span>
        </>
      ),
      content: (
        <>
          <ErrorBoundary alertProps={{ ...defaultErrorBoundaryProps }}>
            <Table<UploadPreviewRow>
              className="margin-bottom-4"
              id={`preview-table-${ageGroup}`}
              rowKey={(row) => row?.name}
              data={ageGroupChildren}
              columns={getPreviewTableColumns()}
              defaultSortColumn={0}
              defaultSortOrder="ascending"
            />
          </ErrorBoundary>
        </>
      ),
      isExpanded: ageGroupChildren.length <= MAX_LENGTH_EXPANDED,
    }));
}

function getChildrenByAgeGroup(
  filteredChildren: UploadPreviewRow[]
): RowsByAgeGroup {
  const childrenByAgeGroup: RowsByAgeGroup = {};
  // Filter all children into an object, keyed by age group
  // with array of children for that agegroup as value
  filteredChildren.reduce((_byAgeGroup, _child) => {
    const ageGroup = _child.ageGroup;
    if (ageGroup) {
      if (!!!_byAgeGroup[ageGroup]) {
        _byAgeGroup[ageGroup] = [_child];
      } else {
        // _byAgeGroup[ageGroup] is not _actually_ possibly undefined; checked in above if
        _byAgeGroup[ageGroup]?.push(_child);
      }
    } else {
      if (!_byAgeGroup[NoAgeGroup]) {
        _byAgeGroup[NoAgeGroup] = [_child];
      } else {
        _byAgeGroup[NoAgeGroup]?.push(_child);
      }
    }
    return _byAgeGroup;
  }, childrenByAgeGroup);

  // Sort object to have consistent section ordering, where NoAgeGroup will always
  // be first if it exists, followed by AgeGroups as they are defined in the enum
  const sortedByAgeGroup: RowsByAgeGroup = {};
  if (
    childrenByAgeGroup[NoAgeGroup] &&
    childrenByAgeGroup[NoAgeGroup]?.length
  ) {
    sortedByAgeGroup[NoAgeGroup] = childrenByAgeGroup[NoAgeGroup];
  }
  Object.values(AgeGroup).forEach((ageGroup) => {
    if (childrenByAgeGroup[ageGroup] && childrenByAgeGroup[ageGroup]?.length) {
      sortedByAgeGroup[ageGroup] = childrenByAgeGroup[ageGroup];
    }
  });
  return sortedByAgeGroup;
}
