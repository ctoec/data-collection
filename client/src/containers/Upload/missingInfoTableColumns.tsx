import { Column } from '@ctoec/component-library';
import React from 'react';
import { EnrollmentColumnError } from '../../shared/payloads';

export const misingInfoTableColumns: () => Column<EnrollmentColumnError>[] = () => {
  const columns: Column<EnrollmentColumnError>[] = [
    {
      name: 'Column name',
      sort: (row) => row.column || '',
      cell: ({ row }) => (
        <th scope="row" className="font-body-2xs">
          <p>{row.formattedName}</p>
        </th>
      ),
    },
    {
      name: '# of errors',
      sort: (row) => row.column || '',
      cell: ({ row }) => (
        <th scope="row" className="font-body-2xs">
          <p>{row.errorCount}</p>
        </th>
      ),
    },
    {
      name: 'Child records with errors',
      sort: (row) => row.column || '',
      cell: ({ row }) => (
        <th scope="row" className="font-body-2xs">
          <p>
            {row.affectedRows.length > 5
              ? row.affectedRows.slice(0, 5).join(', ') +
                ', and ' +
                (row.affectedRows.length - 5).toString() +
                ' more'
              : row.affectedRows.join(', ')}
          </p>
        </th>
      ),
    },
  ];
  return columns;
};
