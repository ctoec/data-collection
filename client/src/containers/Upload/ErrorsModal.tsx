import { Modal, Table } from '@ctoec/component-library';
import React, { Dispatch, SetStateAction } from 'react';
import { Button } from '@ctoec/component-library';
import { ErrorObjectForTable } from './Upload';
import { tableColumns } from './TableColumns';

type ErrorsModalProps = {
  isOpen: boolean;
  toggleIsOpen: () => void;
  clearFile: () => void;
  errorDict: ErrorObjectForTable[];
  setPostUpload: Dispatch<SetStateAction<boolean>>;
};

export const ErrorsModal: React.FC<ErrorsModalProps> = ({
  isOpen,
  toggleIsOpen,
  clearFile,
  errorDict,
  setPostUpload,
}) => {
  return (
    <Modal
      isOpen={isOpen}
      toggleOpen={toggleIsOpen}
      header={<h2> Your upload has missing or incorrect data </h2>}
      showHeaderBorder
      content={
        <div>
          <p>These errors need to be resolved before submitting to OEC.</p>
          <p>
            You can correct records by uploading a corrected file or editing
            records in your ECE Reporter roster.
          </p>

          <Table<ErrorObjectForTable>
            className="margin-bottom-4"
            id="errors-in-sheet-modal-table"
            data={errorDict}
            rowKey={(row) => row.property}
            columns={tableColumns()}
          />

          <div className="margin-top-2">
            <Button
              appearance="outline"
              text="Cancel upload"
              onClick={() => {
                clearFile();
                toggleIsOpen();
              }}
            />
            <Button
              text="Upload and replace in roster"
              onClick={() => {
                setPostUpload(true);
                toggleIsOpen();
              }}
            />
          </div>
        </div>
      }
    />
  );
};
