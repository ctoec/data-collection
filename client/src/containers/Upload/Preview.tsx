import {
  Accordion,
  Button,
  Card,
  HeadingLevel,
  ProgressIndicator,
  ProgressIndicatorProps,
} from '@ctoec/component-library';
import React, { useContext, useEffect, useState } from 'react';
import { useHistory, useLocation } from 'react-router-dom';
import { BackButton } from '../../components/BackButton';
import AuthenticationContext from '../../contexts/AuthenticationContext/AuthenticationContext';
import { BatchUploadResponse } from '../../shared/payloads';
import { apiPost } from '../../utils/api';
import { getFormDataBlob } from '../../utils/getFormDataBlob';
import { getH1RefForTitle } from '../../utils/getH1RefForTitle';
import { handleJWTError } from '../../utils/handleJWTError';
import { clearChildrenCaches } from '../Roster/hooks';
import { getPreviewTableAccordionItems } from './getPreviewTableAccordionItems';

const props: ProgressIndicatorProps = {
  currentIndex: 2,
  steps: [
    {
      label: 'Choose file',
    },
    {
      label: 'Review Missing Info',
    },
    {
      label: 'Preview and upload',
    },
  ],
};

//////////////////////////////////////////////////////

export const Preview: React.FC = () => {
  const h1Ref = getH1RefForTitle();
  const { accessToken } = useContext(AuthenticationContext);
  const history = useHistory();
  const { state } = useLocation();

  //  If there's no state supplied, the user presumably tried to get here manually, which isn't allowed
  if (!state) {
    history.push('/upload');
  }
  const file = (state as any).file;

  const [preview, setPreview] = useState<BatchUploadResponse>();

  useEffect(() => {
    (async function fetchPreview() {
      if (!file) return;

      try {
        const formData = getFormDataBlob(file);
        const resp: BatchUploadResponse = await apiPost(
          `enrollment-reports/false`,
          formData,
          {
            accessToken,
            headers: { 'content-type': formData.type },
            rawBody: true,
          }
        );
        clearChildrenCaches();
        setPreview(resp);
      } catch (error) {
        handleJWTError(history, (err) => {
          console.error(err);
        });
      }
    })();
  }, [file]);

  async function confirmUpload() {
    try {
      const formData = getFormDataBlob(file);
      const resp: BatchUploadResponse = await apiPost(
        `enrollment-reports/true`,
        formData,
        {
          accessToken,
          headers: { 'content-type': formData.type },
          rawBody: true,
        }
      );

      clearChildrenCaches();
      history.push(`/roster`, {
        alerts: [
          {
            type: 'success',
            heading: 'Your records have been uploaded!',
            text: buildUploadSuccessText(resp),
          },
        ],
      });
    } catch (error) {
      handleJWTError(history, (err) => {
        console.error(err);
      });
    }
  }

  function buildUploadSuccessText(batchUpload: BatchUploadResponse) {
    let uploadText = `You uploaded ${batchUpload.new} new records`;
    uploadText +=
      batchUpload.updated > 0 ? `, ${batchUpload.updated} updated records` : '';
    uploadText +=
      batchUpload.withdrawn > 0
        ? ` and ${batchUpload.withdrawn} withdrawn records.`
        : `.`;

    return uploadText;
  }

  const previewAccordionItems = preview
    ? {
        items: getPreviewTableAccordionItems(preview.uploadPreview),
        titleHeadingLevel: 'h2' as HeadingLevel,
      }
    : undefined;

  return (
    <div className="grid-container">
      <BackButton />

      <div className="grid-row display-block">
        <h1 ref={h1Ref} className="margin-bottom-0">
          Upload your enrollment data
        </h1>

        <ProgressIndicator
          currentIndex={props.currentIndex}
          steps={props.steps}
        ></ProgressIndicator>

        <h2 className="margin-bottom-0">
          <span className="usa-step-indicator__current-step">3</span>
          <span className="usa-step-indicator__total-steps"> of 3</span>Preview
          changes and upload file
        </h2>

        <p>
          Here is a summary of the changes you're uploading in this file. If
          everything looks right, upload your changes to your roster.
        </p>
      </div>

      <div className="grid-row desktop:grid-col-4 three-column-card">
        <Card className="font-body-lg">
          <p className="margin-top-0 margin-bottom-0">
            Total records in this file
          </p>
          <p className="text-bold margin-top-0 margin-bottom-0">
            {preview ? preview.new + preview.updated + preview.withdrawn : ''}
          </p>
        </Card>
      </div>
      <div className="grid-row three-column-layout">
        <div className="desktop:grid-col-4 three-column-card">
          <Card className="font-body-lg">
            <p className="margin-top-0 margin-bottom-0">New</p>
            <p className="text-bold margin-top-0 margin-bottom-0">
              {preview ? preview.new : ''}
            </p>
          </Card>
        </div>
        <div className="desktop:grid-col-4 three-column-card">
          <Card className="font-body-lg">
            <p className="margin-top-0 margin-bottom-0">Updated</p>
            <p className="text-bold margin-top-0 margin-bottom-0">
              {preview ? preview.updated : ''}
            </p>
          </Card>
        </div>
        <div className="desktop:grid-col-4 three-column-card">
          <Card className="font-body-lg">
            <p className="margin-top-0 margin-bottom-0">Withdrawn</p>
            <p className="text-bold margin-top-0 margin-bottom-0">
              {preview ? preview.withdrawn : ''}
            </p>
          </Card>
        </div>
      </div>

      <div className="grid-row">
        {previewAccordionItems ? (
          <Accordion {...previewAccordionItems} />
        ) : (
          <></>
        )}
      </div>

      <Button text="Cancel Upload" href="/home" appearance="outline" />
      <Button
        id="upload-button"
        text="Save changes to roster"
        onClick={confirmUpload}
      />
    </div>
  );
};
