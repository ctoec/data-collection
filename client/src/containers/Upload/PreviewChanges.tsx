import {
  Accordion,
  Button,
  Card,
  LoadingWrapper,
} from '@ctoec/component-library';
import React, { useContext, useState } from 'react';
import { useHistory } from 'react-router-dom';
import { stringify } from 'query-string';
import AuthenticationContext from '../../contexts/AuthenticationContext/AuthenticationContext';
import RosterContext from '../../contexts/RosterContext/RosterContext';
import { EnrollmentReportUploadResponse } from '../../shared/payloads';
import { FixedBottomBar } from '../../components/FixedBottomBar/FixedBottomBar';
import { useAlerts } from '../../hooks/useAlerts';
import { apiPost } from '../../utils/api';
import { handleJWTError } from '../../utils/handleJWTError';
import { getPreviewTableAccordionItems } from './getPreviewTableAccordionItems';
import { StepContentProps } from './UploadWizard';

export const PreviewChanges: React.FC<StepContentProps> = ({
  enrollmentReportCheck,
  setCurrentStepIndex,
  setEnrollmentReportCheck,
}) => {
  if (!enrollmentReportCheck) throw new Error('Something went wrong');

  const {
    childRecords,
    counts: { newCount, updatedCount, withdrawnCount, unchangedCount },
  } = enrollmentReportCheck;

  const history = useHistory();
  const { accessToken } = useContext(AuthenticationContext);
  const { query, revalidate } = useContext(RosterContext);
  const [_, setAlerts] = useAlerts();

  const [loading, setLoading] = useState(false);
  const [loadingText, setLoadingText] = useState('');

  async function confirmUpload() {
    setLoadingText('Uploading changes...');
    setLoading(true);

    try {
      const resp: EnrollmentReportUploadResponse = await apiPost(
        `enrollment-reports`,
        { childRecords },
        { accessToken }
      );

      await revalidate();
      history.push({
        pathname: '/roster',
        search: stringify(query),
        state: {
          alerts: [
            {
              type: 'success',
              heading: 'Your records have been uploaded!',
              text: buildUploadSuccessText(resp),
            },
          ],
        },
      });
    } catch (error) {
      handleJWTError(history, (err) => {
        console.error(err);
        setAlerts([
          {
            type: 'error',
            text: 'Something went wrong with your upload!',
          },
        ]);
      })(error);
    } finally {
      setLoading(false);
    }
  }

  function buildUploadSuccessText(
    uploadResponse: EnrollmentReportUploadResponse
  ) {
    let uploadText = `You uploaded ${uploadResponse.newCount} new records`;
    uploadText +=
      uploadResponse.updatedCount > 0
        ? `, ${uploadResponse.updatedCount} updated records`
        : '';
    uploadText +=
      uploadResponse.withdrawnCount > 0
        ? ` and ${uploadResponse.withdrawnCount} withdrawn records.`
        : `.`;

    return uploadText;
  }

  return (
    <>
      <p>
        Here is a summary of the changes you're uploading in this file. If
        everything looks right, upload your changes to your roster.
      </p>
      <LoadingWrapper text={loadingText} loading={loading}>
        <>
          <div className="grid-container upload-roster">
            <div className="grid-row desktop:grid-col-4 three-column-card">
              <Card className="font-body-lg">
                <p className="margin-top-0 margin-bottom-0">
                  Total records in this file
                </p>
                <p className="text-bold margin-top-0 margin-bottom-0">
                  {newCount + updatedCount + withdrawnCount + unchangedCount}
                </p>
              </Card>
            </div>
            <div className="grid-row three-column-layout">
              <div className="desktop:grid-col-4 three-column-card">
                <Card className="font-body-lg">
                  <p className="margin-top-0 margin-bottom-0">New</p>
                  <p className="text-bold margin-top-0 margin-bottom-0">
                    {newCount}
                  </p>
                </Card>
              </div>
              <div className="desktop:grid-col-4 three-column-card">
                <Card className="font-body-lg">
                  <p className="margin-top-0 margin-bottom-0">Updated</p>
                  <p className="text-bold margin-top-0 margin-bottom-0">
                    {updatedCount}
                  </p>
                </Card>
              </div>
              <div className="desktop:grid-col-4 three-column-card">
                <Card className="font-body-lg">
                  <p className="margin-top-0 margin-bottom-0">Withdrawn</p>
                  <p className="text-bold margin-top-0 margin-bottom-0">
                    {withdrawnCount}
                  </p>
                </Card>
              </div>
            </div>

            <div className="grid-row upload-preview-table">
              <Accordion
                items={getPreviewTableAccordionItems(childRecords)}
                titleHeadingLevel="h3"
              />
            </div>
          </div>
          <FixedBottomBar>
            <Button
              text="Cancel upload"
              appearance="outline"
              onClick={() => {
                setEnrollmentReportCheck(undefined);
                setCurrentStepIndex(0);
              }}
            />
            <Button
              id="upload-button"
              text="Save changes to roster"
              onClick={confirmUpload}
            />
          </FixedBottomBar>
        </>
      </LoadingWrapper>
    </>
  );
};
