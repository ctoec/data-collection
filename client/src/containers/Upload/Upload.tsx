import React, { useState, useContext, useEffect } from 'react';
import cx from 'classnames';
import { Link, useHistory } from 'react-router-dom';
import AuthenticationContext from '../../contexts/AuthenticationContext/AuthenticationContext';
import {
  FileInput,
  TextWithIcon,
  Alert,
  LoadingWrapper,
  ErrorBoundary,
} from '@ctoec/component-library';
import { ReactComponent as Arrow } from '@ctoec/component-library/dist/assets/images/arrowRight.svg';
import { apiPost, apiGet } from '../../utils/api';
import { getErrorHeading, getErrorText } from '../../utils/error';
import { getH1RefForTitle } from '../../utils/getH1RefForTitle';
import { handleJWTError } from '../../utils/handleJWTError';
import { CheckReplaceData } from './CheckReplaceData';
import { CSVExcelDownloadButton } from '../../components/CSVExcelDownloadButton';
import { ErrorModal } from './ErrorModal/ErrorsModal';
import { ErrorObjectForTable } from './ErrorModal/ErrorObjectForTable';
import { clearChildrenCaches } from '../Roster/hooks';
import { defaultErrorBoundaryProps } from '../../utils/defaultErrorBoundaryProps';
import { BatchUpload } from '../../shared/payloads';

const Upload: React.FC = () => {
  const h1Ref = getH1RefForTitle();
  const { accessToken } = useContext(AuthenticationContext);
  const history = useHistory();

  // Count how many children are in the roster so we can determine if we're writing over that data
  const [userRosterCount, setUserRosterCount] = useState(undefined);
  useEffect(() => {
    apiGet('children?count=true', accessToken)
      .then((res) => setUserRosterCount(res.count))
      .catch((err) => {
        throw new Error(err);
      });
  }, [accessToken]);

  const [error, setError] = useState<string>();
  const [loading, setLoading] = useState(false);

  // Check the file for errors if there is a file
  const [file, setFile] = useState<File>();
  const [errorDict, setErrorDict] = useState<ErrorObjectForTable[]>();
  useEffect(() => {
    // Haven't yet determined how many errors of each type there are
    if (file && errorDict === undefined) {
      setLoading(true);
      // Ugh internet explorer why
      // https://developer.mozilla.org/en-US/docs/Web/API/FormData
      // https://www.npmjs.com/package/formdata-polyfill
      const formData = new FormData();
      formData.append('file', file, file.name);
      apiPost(`enrollment-reports/check`, formData, {
        accessToken,
        rawBody: true,
      })
        // Back end sends back an object whose fields are error table obj.
        .then((resp) => {
          setErrorDict(resp);
        })
        .catch(
          handleJWTError(history, (err) => {
            setError(err);
            clearFile();
            setErrorDict(undefined);
          })
        )
        .finally(() => setLoading(false));
    }
  }, [file, errorDict]);

  // If the file exists and the upload should be posted,
  // then trigger the API request
  const [postUpload, setPostUpload] = useState(false);
  const [queryStringForUpload, setQueryStringForUpload] = useState('');
  useEffect(() => {
    if (file && postUpload) {
      setLoading(true);
      const formData = new FormData();
      formData.set('file', file);
      apiPost(`enrollment-reports${queryStringForUpload}`, formData, {
        accessToken,
        rawBody: true,
      })
        // Response contains id of created enrollmentReport,
        // number of active enrollments, and num withdrawn enrollments
        // via BatchUpload payload
        .then((resp: BatchUpload) => {
          // Clear all children records from data cache
          clearChildrenCaches();
          let uploadText = `You uploaded ${resp.activeEnrollments} active enrollments`;
          uploadText +=
            resp.withdrawnEnrollments > 0
              ? ` and ${resp.withdrawnEnrollments} withdrawn enrollments.`
              : `.`;
          history.push(`/roster`, {
            alerts: [
              {
                type: 'success',
                heading: 'Your records have been uploaded!',
                text: uploadText,
              },
            ],
          });
        })
        .catch(
          handleJWTError(history, (err) => {
            setError(err);
            clearFile();
          })
        )
        // Reset this flag to false so the upload can be subsequently re-triggered
        .finally(() => {
          setPostUpload(false);
          setLoading(false);
        });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [postUpload]);

  // Case where the user doesn't already have a roster so we
  // don't need to go to CheckReplace: next step is confirm upload
  const [errorModalOpen, setErrorModalOpen] = useState(false);
  const advanceToPostUpload = () => {
    setPostUpload(true);
    setErrorModalOpen(false);
  };

  // Case where the user does have an existing roster; need to
  // open the CheckReplace modal as a next step
  const [checkReplaceDataOpen, setCheckReplaceDataOpen] = useState(false);
  const advanceToCheckReplace = () => {
    setErrorModalOpen(false);
    setCheckReplaceDataOpen(true);
  };
  useEffect(() => {
    // wait until we know if the user already has a roster or not
    // before allowing them to submit data
    if (userRosterCount === undefined) return;

    // If they have selected a file, then open the error checking modal.
    // If they confirm the upload, decide if we just post the upload
    // or we display check replace data modal
    if (file && errorDict !== undefined) {
      if (userRosterCount === 0) {
        // If state has an empty list, back-end found no errors
        if (errorDict.length > 0) setErrorModalOpen(true);
        else setPostUpload(true);
      } else {
        if (errorDict.length > 0) setErrorModalOpen(true);
        else setCheckReplaceDataOpen(true);
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [file, userRosterCount, errorDict]);

  const [fileKey, setFileKey] = useState(0);
  const clearFile = () => {
    // When the file is cleared, change the key to force the file component to rerender/reset
    setFile(undefined);
    setFileKey((oldKey) => oldKey + 1);
  };
  const fileUpload = (e: React.ChangeEvent<HTMLInputElement>) => {
    e.preventDefault();
    const _file = e?.target?.files?.[0];
    if (!_file) {
      clearFile();
      return setError('No file selected for upload');
    }
    setFile(_file);
    setError(undefined);
  };

  return (
    <div className="grid-container margin-top-4">
      <ErrorModal
        isOpen={errorModalOpen}
        toggleIsOpen={() => setErrorModalOpen((o) => !o)}
        clearFile={() => {
          clearFile();
          setErrorDict(undefined);
        }}
        errorDict={errorDict || []}
        nextFunc={
          userRosterCount === 0 ? advanceToPostUpload : advanceToCheckReplace
        }
      />
      <CheckReplaceData
        isOpen={checkReplaceDataOpen}
        clearFile={clearFile}
        toggleIsOpen={() => setCheckReplaceDataOpen((o) => !o)}
        setPostUpload={setPostUpload}
        setQueryString={setQueryStringForUpload}
      />
      {error && (
        <div className="margin-bottom-2">
          <Alert
            heading={getErrorHeading(error)}
            text={getErrorText(error)}
            type="error"
            actionItem={
              <div>
                <p className="margin-bottom-1 text-bold">
                  Download the data collection template
                </p>
                <CSVExcelDownloadButton
                  fileType="xlsx"
                  whichDownload="template"
                  className="margin-bottom-1"
                />
                <CSVExcelDownloadButton
                  fileType="csv"
                  whichDownload="template"
                  className="margin-bottom-1"
                />
              </div>
            }
          />
        </div>
      )}
      <div className="margin-bottom-2 text-bold">
        <Link className="usa-button usa-button--unstyled" to="/">
          <TextWithIcon
            text="Back"
            Icon={Arrow}
            direction="left"
            iconSide="left"
          />
        </Link>
      </div>

      <div className="grid-row">
        <h1 ref={h1Ref}>Upload your enrollment data</h1>
        <p>
          After you've entered all state funded enrollment data in the
          spreadsheet template, upload the file here.
        </p>
      </div>
      <ErrorBoundary alertProps={{ ...defaultErrorBoundaryProps }}>
        <div className="grid-row">
          <form
            className={cx('usa-form', {
              'display-none': checkReplaceDataOpen || errorModalOpen,
            })}
          >
            <LoadingWrapper text="Uploading your file..." loading={loading}>
              <FileInput
                key={fileKey}
                id="report"
                label="Choose a file"
                onChange={fileUpload}
              />
            </LoadingWrapper>
          </form>
        </div>
      </ErrorBoundary>
    </div>
  );
};

export default Upload;
