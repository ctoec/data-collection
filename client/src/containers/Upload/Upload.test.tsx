import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { snapshotTestHelper, accessibilityTestHelper } from '../../testHelpers';
import Upload from './Upload';

const routerWrapped = (
  <BrowserRouter>
    <Upload />
  </BrowserRouter>
);
describe('Upload', () => {
  snapshotTestHelper(routerWrapped);
  accessibilityTestHelper(routerWrapped);
});
