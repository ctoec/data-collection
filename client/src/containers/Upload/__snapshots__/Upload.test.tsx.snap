// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid-container margin-top-4"
  >
    <div
      class="margin-bottom-2 text-bold"
    >
      <a
        href="/"
      >
        <span
          class="oec-text-with-icon"
        >
          <svg
            class="oec-text-with-icon__icon oec-text-with-icon__icon--direction-left oec-text-with-icon__icon--side-left"
          >
            arrowRight.svg
          </svg>
          Back
        </span>
      </a>
    </div>
    <div
      class="grid-row"
    >
      <h1>
        Upload your enrollment data
      </h1>
      <p>
        After you've entered all state funded enrollment data in the spreadsheet template, upload the file here.
      </p>
    </div>
    <div
      class="grid-row"
    >
      <form
        class="usa-form"
      >
        <label
          for="report"
        >
          Upload enrollment data
        </label>
        <input
          class="usa-file-input"
          id="report"
          type="file"
        />
      </form>
    </div>
  </div>
</DocumentFragment>
`;
